api/src/main/java/org/pmiops/workbench/cohortbuilder/FieldSetQueryBuilder.java

```diff
@@ -42,14 +52,189 @@
+    private Map<String, String> aliasToTableName = new HashMap<>();
+    // The columns selected in the select clause.
+    private ImmutableList.Builder<ColumnInfo> selectColumns = ImmutableList.builder();
+    // The query being turned into SQL.
+    private TableQuery tableQuery;
+    // The CDR schema configuration.
+    private CdrBigQuerySchemaConfig schemaConfig;
+    // A map of (column name -> ColumnConfig) for the columns on the main table (pulled from
+    // columnConfigTable).
+    private Map<String, ColumnConfig> mainTableColumns;
+  }
+
+  private static final class TableNameAndAlias {
+    private final String tableName;
+    private final String alias;
+
+    TableNameAndAlias(String tableName, String alias) {
+      this.tableName = tableName;
+      this.alias = alias;
+    }
+  }
+
   @Autowired
   public FieldSetQueryBuilder(ParticipantCounter participantCounter) {
     this.participantCounter = participantCounter;
   }
 
-  private void handleComparison(ColumnFilter columnFilter, ColumnConfig columnConfig,
-      StringBuilder sqlBuilder, Map<String, QueryParameterValue> paramMap) {
-    String paramName = "p" + paramMap.size();
+  private ColumnConfig findPrimaryKey(Iterable<ColumnConfig> columnConfigs) {
+    for (ColumnConfig columnConfig : columnConfigs) {
+      if (columnConfig.primaryKey != null && columnConfig.primaryKey) {
+        return columnConfig;
+      }
+    }
+    throw new IllegalStateException("Table lacks primary key!");
+  }
+
+  private List<String> parseColumnName(String columnName) {
+    return Splitter.on(TABLE_SEPARATOR).splitToList(columnName);
+  }
+
+  private String toTableAlias(List<String> columnParts, int endIndex) {
+    return Joiner.on(ALIAS_SEPARATOR).join(columnParts.subList(0, endIndex));
+  }
+
+  private String getForeignKeyColumn(String columnPart) {
+    return columnPart + "_id";
+  }
+
+  private Map<String, ColumnConfig> getColumnConfigs(
+      QueryState queryState, String tableName, boolean cohortTable) {
+    Map<String, ColumnConfig> configTable = queryState.columnConfigTable.get(tableName);
+    if (configTable == null) {
+      TableConfig tableConfig = queryState.schemaConfig.cohortTables.get(tableName);
+      if (tableConfig == null) {
+        if (cohortTable) {
+          throw new BadRequestException("Not a valid cohort table: " + tableName);
+        }
+        tableConfig = queryState.schemaConfig.metadataTables.get(tableName);
+        if (tableConfig == null) {
+          throw new BadRequestException("Table not found: " + tableName);
+        }
+      }
+      configTable = new HashMap<>();
+      queryState.columnConfigTable.put(tableName, configTable);
+      for (ColumnConfig columnConfig : tableConfig.columns) {
+        configTable.put(columnConfig.name, columnConfig);
+      }
+    }
+    return configTable;
+  }
+
+  private TableNameAndAlias getTableNameAndAlias(List<String> columnParts, QueryState queryState) {
```

Comment:
It took me a very long time to grok this method. I'm not sure it can be significantly improved for readability, but I think for me this is reinforcing a need to have unit tests for this class (something with faster iteration and higher run frequency than bq integration tests).

Even something that just made assertions about the `QueryState` might go a long way.